File: year2023/day01/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const digitMap = {'one':'1','two':'2','three':'3','four':'4','five':'5','six':'6','seven':'7','eight':'8','nine':'9'};
const input: string[] = file.toString().split('\n');

if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return input.reduce((acc, val) => {
    const digits = Array.from(val).filter(char => /\d/.test(char));
    return acc + Number(digits[0] + digits[digits.length - 1]);
  }, 0);
}

export function part_2(): number {
  const mapDigit = input => digitMap.hasOwnProperty(input) ? digitMap[input] : input;
  return input.reduce((acc, val) => {
    const digits = val.match(/(?:1|2|3|4|5|6|7|8|9|one|two|three|four|five|six|seven|eight|nine)/g).map(item => mapDigit(item));
    return acc + Number(digits[0] + digits[digits.length-1]);
  }, 0);
}

File: year2023/day01/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  sum = 0
  for val in input:
    sum += int(next((char for char in val if char.isdigit()), None) + next((char for char in reversed(val) if char.isdigit()), None))

  return sum


def part_2():
  sum = 0
  for val in input:
    digits = []

    for i, char in enumerate(val):
      if char.isdigit():
          digits.append(char)

      for digit, str_num in enumerate(['zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']):
        if val[i:].startswith(str_num):
          digits.append(str(digit))

    sum += int(digits[0] + digits[-1])
    
  return sum


def setup(path):
  global input

  with open(path, 'r') as file:
    input = [row for row in file]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()